river/LayoutManager.zig

```diff
@@ -0,0 +1,378 @@
+                // is raised, so that the client then has no idea that its previously
+                // perfectly fine serial is now invalid. So we just ignore requests with
+                // wrong serials.
+                if (self.output.layout_demand == null) return;
+                if (self.output.layout_demand.?.serial != req.serial) return;
+
+                self.output.layout_demand.?.pushViewDimensions(req.x, req.y, req.width, req.height) catch {};
+            },
+
+            // We receive this event when the client wants to mark the proposed layout
+            // of the layout demand matching the serial as done.
+            .commit => |req| {
+                log.debug(
+                    .layout_manager,
+                    "layout '{}' on output '{}' commited",
+                    .{ self.namespace, self.output.wlr_output.name },
+                );
+
+                // Are we a valid layout?
+                if (self.namespace == null) return;
+
+                // Does the serial match?
+                if (self.output.layout_demand == null) return;
+                if (self.output.layout_demand.?.serial != req.serial) return;
+
+                // From now on, no matter what happens, we want to destroy the layout demand
+                // and start a transaction.
+                defer {
+                    self.output.layout_demand.?.deinit();
+                    util.gpa.destroy(self.output.layout_demand.?);
+                    self.output.layout_demand = null;
+                    self.output.root.startTransaction();
+                }
+
+                // Check if amount of proposed dimensions matches amount of views to arrange
+                if (self.output.layout_demand.?.views != self.output.layout_demand.?.view_dimensions.len) {
+                    log.err(.layout_manager, "amount of proposed view dimensions does not match amount of views; Aborting layout demand", .{});
+                    layout.postError(
+                        .proposed_dimension_mismatch,
+                        "the amount of proposed view dimensions needs to match the amount of views",
+                    );
+                    return;
+                }
+
+                // Apply proposed layout to views and start transaction
+                var it = ViewStack(View).iter(self.output.views.first, .forward, self.output.pending.tags, Output.arrangeFilter);
+                while (it.next()) |view| {
+                    view.pending.box = self.output.layout_demand.?.popViewDimensions() orelse unreachable;
+                    view.applyConstraints();
+                }
+                self.output.pending.layout = true;
+
+                // TODO layout name?
+            },
+        }
+    }
+};
+
+/// This object encapsulates a zriver.LayoutManagerV1 object bound by a client
+const LayoutManager = struct {
+    parent: *Self,
+    layout_manager: *zriver.LayoutManagerV1,
```

Comment:
I'm not sure splitting this off into another struct is really worth it. The implementations of river-control and river-status simply do everything in the toplevel object. I don't think you even need to store a list of `*zriver.LayoutManagerV1` pointers.